lib/mss/content_protection.js

```diff
@@ -0,0 +1,338 @@
+    const view = shaka.util.BufferUtils.toDataView(data);
+
+    // First 4 bytes is the PRO length (DWORD)
+    const byteLength = view.getUint32(byteOffset, /* littleEndian= */ true);
+    byteOffset += 4;
+
+    if (byteLength != data.byteLength) {
+      // Malformed PRO
+      shaka.log.warning('PlayReady Object with invalid length encountered.');
+      return [];
+    }
+
+    // Skip PRO Record count (WORD)
+    byteOffset += 2;
+
+    // Rest of the data contains the PRO Records
+    const ContentProtection = shaka.mss.ContentProtection;
+    return ContentProtection.parseMsProRecords_(view, byteOffset);
+  }
+
+  /**
+   * PlayReady Header format: https://goo.gl/dBzxNA
+   *
+   * @param {!Element} xml
+   * @return {string}
+   * @private
+   */
+  static getLaurl_(xml) {
+    // LA_URL element is optional and no more than one is
+    // allowed inside the DATA element. Only absolute URLs are allowed.
+    // If the LA_URL element exists, it must not be empty.
+    for (const elem of xml.getElementsByTagName('DATA')) {
+      for (const child of elem.childNodes) {
+        if (child instanceof Element && child.tagName == 'LA_URL') {
+          return child.textContent;
+        }
+      }
+    }
+
+    // Not found
+    return '';
+  }
+
+  /**
+   * Gets a PlayReady license URL from a protection element
+   * containing a PlayReady Header Object
+   *
+   * @param {!Element} element
+   * @return {string}
+   */
+  static getPlayReadyLicenseUrl(element) {
+    const ContentProtection = shaka.mss.ContentProtection;
+    const PLAYREADY_RECORD_TYPES = ContentProtection.PLAYREADY_RECORD_TYPES;
+
+    const bytes = shaka.util.Uint8ArrayUtils.fromBase64(element.textContent);
+    const records = ContentProtection.parseMsPro_(bytes);
+    const record = records.filter((record) => {
+      return record.type === PLAYREADY_RECORD_TYPES.RIGHTS_MANAGEMENT;
+    })[0];
+
+    if (!record) {
+      return '';
+    }
+
+    const xml = shaka.util.StringUtils.fromUTF16(record.value, true);
+    const rootElement = shaka.util.XmlUtils.parseXmlString(xml, 'WRMHEADER');
+    if (!rootElement) {
+      return '';
+    }
+
+    return ContentProtection.getLaurl_(rootElement);
+  }
+
+  /**
+   * PlayReady Header format: https://goo.gl/dBzxNA
+   *
+   * @param {!Element} xml
+   * @return {?string}
+   * @private
+   */
+  static getKID_(xml) {
+    // KID element is optional and no more than one is
+    // allowed inside the DATA element.
+    for (const elem of xml.getElementsByTagName('DATA')) {
+      for (const child of elem.childNodes) {
+        if (child instanceof Element && child.tagName == 'KID') {
+          return child.textContent;
+        }
+      }
+    }
+
+    // Not found
+    return null;
+  }
+
+  /**
+   * Gets a PlayReady KID from a protection element
+   * containing a PlayReady Header Object
+   *
+   * @param {!Element} element
+   * @return {?string}
+   * @private
+   */
+  static getPlayReadyKID_(element) {
```

Comment:
This function is almost exactly the same as `getPlayReadyLicenseUrl`.
You could write a utility function that gets the root element of a `RIGHTS_MANAGEMENT` node, and make these two functions like three lines each.

Reply:
Added a new method